### PR TITLE
Add ZenGRC Ingestion Script

### PR DESCRIPTION
**Pull request by greg-snowflake**
To test: Add the environment variables defined at the top of the file, then run `python zengrc.py`.

Code was tested in our DEV environment and confirmed that data ingests when scheduled, and file exits cleanly when ingestion is not scheduled.

### CODE REVIEW

File: src/ingestion/zengrc.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+from runners.helpers import db, log
+
+ZENGRC_ID = os.environ.get('ZENGRC_ID')
+ZENGRC_SECRET = os.environ.get('ZENGRC_SECRET')
+ZENGRC_URL = os.environ.get('ZENGRC_URL')
+ZENGRC_TABLE = os.environ.get('ZENGRC_TABLE')
+TIMESTAMP = str(datetime.datetime.utcnow())
+
+
+def gen_headers(id, secret):
```
Comment (andrey-snowflake) [inline src/ingestion/zengrc.py]: please use the auth param like we do in auth.py — https://github.com/snowflakedb/SnowAlert/blob/master/src/runners/helpers/auth.py#L42

File: src/ingestion/zengrc.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+
+def loop(endpoint):
+    log.info(f"starting {endpoint}")
+    j = {'links': {'next': {'href': endpoint}}}
+    headers = gen_headers(ZENGRC_ID, ZENGRC_SECRET)
+    page = 1
+    while j['links']['next'] is not None:
+        log.info(f"Getting page {str(page)}")
+        r = requests.get(ZENGRC_URL+j['links']['next']['href'], headers=headers)
```
Comment (andrey-snowflake) [inline src/ingestion/zengrc.py]: please abstract out into a separate `get(path)` method that prepends ZENGRC_URL and sets auth headers

File: src/ingestion/zengrc.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+
+def gen_headers(id, secret):
+    auth = base64.b64encode(f'{id}:{secret}'.encode()).decode()
+    return {"Authorization": f"Basic {auth}"}
+
+
+def loop(endpoint):
+    log.info(f"starting {endpoint}")
+    j = {'links': {'next': {'href': endpoint}}}
```
Comment (andrey-snowflake) [inline src/ingestion/zengrc.py]: what is `j`? is there a better name for it?

File: src/ingestion/zengrc.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+ZENGRC_TABLE = os.environ.get('ZENGRC_TABLE')
+TIMESTAMP = str(datetime.datetime.utcnow())
+
+
+def gen_headers(id, secret):
+    auth = base64.b64encode(f'{id}:{secret}'.encode()).decode()
+    return {"Authorization": f"Basic {auth}"}
+
+
+def loop(endpoint):
```
Comment (andrey-snowflake) [inline src/ingestion/zengrc.py]: I would expect a function called `loop` to be really general, but here it seems relatively specific to the fact that you're using it to implement looping through pagination?. maybe `update_data` could work? in general functions should be named after the side-effects they have, where looping is an implementation detail.

File: src/ingestion/zengrc.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        if r.status_code != 200:
+            log.error(f"Ingest request for {endpoint} failed", r.text)
+            db.ingest_request_failed(ZENGRC_TABLE, r, TIMESTAMP)
+            break
+
+        j = json.loads(r.text)
+        data = [[json.dumps(i), TIMESTAMP] for i in j['data']]
+        if len(data) > 0:
+            try:
+                query = f"INSERT INTO {ZENGRC_TABLE} select parse_json(column1), column2 from values "
```
Comment (andrey-snowflake) [inline src/ingestion/zengrc.py]: pushed a minor cleanup here -- can you give me keys to test it, or test?

File: src/ingestion/zengrc.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+        if r.status_code != 200:
+            log.error(f"Ingest request for {endpoint} failed", r.text)
+            db.ingest_request_failed(ZENGRC_TABLE, r, TIMESTAMP)
+            break
+
+        j = json.loads(r.text)
+        data = [[json.dumps(i), TIMESTAMP] for i in j['data']]
+        try:
+            db.insert(ZENGRC_TABLE, data, select='PARSE_JSON(column1), column2')
```
Comment (andrey-snowflake) [inline src/ingestion/zengrc.py]: changed this -- can you give me keys to test or test yourself?

File: src/runners/helpers/db.py [hunk 7]
```diff
@@ -344,3 +349,13 @@ def record_metadata(metadata, table, e=None):
 
     except Exception as e:
         log.error(f"{record_type} metadata failed to log.", e)
+
+
+def ingest_request_failed(table, r, timestamp):
```
Comment (andrey-snowflake) [inline src/runners/helpers/db.py]: functions should be called after what they do, and events (in pub/sub model) after what it is that happened. how about `record_failed_ingestion` here?

File: src/runners/helpers/db.py [hunk 8]
```diff
@@ -344,3 +349,13 @@ def record_metadata(metadata, table, e=None):
...
+    log = json.dumps({'headers': dict(r.headers),
+                      'text': r.text,
+                      'status_code': r.status_code,
+                      'failure': True})
+    data = [(log, timestamp)]
+    query = f"INSERT INTO {table} SELECT PARSE_JSON(COLUMN1), COLUMN2 FROM VALUES (%s)"
```
Comment (andrey-snowflake) [inline src/runners/helpers/db.py]: reuse db.insert?

File: src/ingestion/zengrc.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+def main():
+    print("starting")
+    endpoints = ['/api/v2/assessments',
+                 '/api/v2/audits',
+                 '/api/v2/issues',
+                 '/api/v2/requests',
+                 '/api/v2/risks']
+
+    # 1435 is five minutes short of one day, to prevent slow drift of ingest time.
```
[the remark this answers is not in the document]
Comment (greg-snowflake) [reply src/ingestion/zengrc.py]: We check every fifteen minutes to see if the data is over some threshold of stale. You can imagine a scenario in which when that check happens, the data is a few seconds shy of being stale, and so ingest is delayed for fifteen minutes. 24 hours later, we check and data is a few seconds shy of being stale, so ingest is delayed for fifteen minutes (which is now 30 minutes past when it was ingested two days ago).

File: src/runners/helpers/db.py (hunk 8 above)
Comment (greg-snowflake) [reply src/runners/helpers/db.py]: db.insert doesn't generate the correct sql.

File: src/ingestion/zengrc.py (hunk 1 above)
Comment (greg-snowflake) [reply src/ingestion/zengrc.py]: done

File: src/runners/helpers/db.py (hunk 7 above)
Comment (greg-snowflake) [reply src/runners/helpers/db.py]: done

File: src/ingestion/zengrc.py (hunk 2 above)
Comment (greg-snowflake) [reply src/ingestion/zengrc.py]: done

File: src/ingestion/zengrc.py (hunk 3 above)
Comment (greg-snowflake) [reply src/ingestion/zengrc.py]: resolved

File: src/ingestion/zengrc.py (hunk 4 above)
Comment (greg-snowflake) [reply src/ingestion/zengrc.py]: resolved

File: src/ingestion/zengrc.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (andrey-snowflake) [reply src/ingestion/zengrc.py]: ah, got it. can update to make more clear

File: src/runners/helpers/db.py (hunk 8 above)
Comment (andrey-snowflake) [reply src/runners/helpers/db.py]: I mean, tests pass ;)

**Pull request merged**